node/src/Unison/Runtime/ResourcePool.hs

```diff
@@ -0,0 +1,64 @@
+module Unison.Runtime.ResourcePool where
+
+import qualified Data.Map as M
+import qualified Control.Concurrent.MVar as MVar
+import Data.Time (UTCTime, getCurrentTime, addUTCTime)
+import qualified Control.Concurrent as CC
+
+-- acquire returns the resource, and the cleanup action ("finalizer") for that resource
+data Pool p r = Pool { acquire :: p -> Int -> IO (r, IO ()) }
+
+type ResourceKey p = (p,CC.ThreadId)
+type Cache p r = MVar.MVar (M.Map (ResourceKey p) (r, UTCTime, IO ()))
+
+addResourceToMap :: (Ord p) =>  (r -> IO()) -> Cache p r -> p -> r -> Int -> Int -> IO(CC.ThreadId) -> IO ()
```

Comment:
Minor formatting stuff -

* If just one constraint, leave off the parens - `Ord p => ...`.
* `IO ()` and `IO CC.ThreadId`, type application should be formatted same as function application - space after the type constructor, no parens around the args unless necessary